Parser Fixes and Coldbox RESTful App Tests

homestar9 opened the pull request:
I spent a lot of time reviewing the Parser to figure out what was happening in the last update. I have isolated the problem, fixed it, made some improvements, and added a performance boost.

I also created a new test spec called ColdboxSpec, designed to test common use cases where Coldbox RESTful applications need to generate OpenApi (swagger) files.  Coldbox API's use different conventions than PetStore, and testing against these conventions can help prevent future issues and optimize development.

There is one breaking change in this patch compared with the development branch.  I consolidated the `$allOf` and `$oneOff` composition keys into a single key called `$extend`.   `$extend` is a better descriptor of how the process works because it _extends_ multiple JSON files and flattens them into a single object (similar to CFML's `structAppend()` or jQuery's `extend()`).  The terms `allOf` and `oneOf` were misleading because they are used by swagger files to denote _inheritance_ and _polymorphism_.  According to the OpenAPI specs, `allOf` can only be used in the `components.schemas` section of the swagger file.   This change should help avoid confusion moving forward.

## ColdBox Tests Spec:

Coldbox RESTful apps use a set of conventions to build out APIs, including a powerful base REST handler.  I felt it was important to build a test spec to emulate and check against a simulated typical Coldbox REST API's request and response structure.  The primary goals of the new test specs were to:
 - Optimize for model/object reuse across the docs
 - Minimize copy/pasting of boilerplate JSON data
 - Use a typical Coldbox REST handler naming convention to create common routes/endpoints
 - Utilize a combination of recursive `$ref` injections and internal references to account for coding preferences.
 - Utilize the new `$extend` key to build leaner swagger files.

I added a new folder to the test-harness/tests/resources/ folder called "coldboxRest". The structure should emulate a typical ColdBox app, with schemas, handlers (paths), responses, and examples.

I immediately noticed how clean the handler (path) JSON files became once I started using `$ref` annotations to point to internal and external references. Here's an example:
```
// posts.json
{
    "/posts": {
        "get": {
            "summary": "Paginate posts",
            "operationId": "posts.index",
            "tags": ["posts"],
            "parameters": {
                "$ref": "index/parameters.json#/parameters"
            },
            "responses": {
                "$ref": "index/responses.json"
            }
        } 
    },
    "/posts/{id}": {
        "get": {
            "summary": "Show a post",
            "operationId": "posts.show",
            "tags": ["posts"],
            "parameters": {
                "$ref": "show/parameters.json#/parameters"
            },
            "responses": {
                "$ref": "show/responses.json"
            }
        }
    }
}
```

I also accounted for different coding styles with `$ref` keys.  Some people inject every model directly into the response schema, whereas others prefer referencing models in the `components.schemas` section using the new double hash `##` annotation.

Here's an example of how both work:
```
// /paths/users/index/responses.json
{
  "200": {
    "description": "Success",
    "content": {
      "application/json": {
        "schema": {
          "$extend": [
            {
              "type": "object"
            },
            {
              "$ref": "../../../_schemas/response.json" <- injection
            },
            {
              "properties": {
                "data": {
                  "type": "object",
                  "properties": {
                    "results": {
                      "type": "array",
                      "items": {
                        "$ref": "../_schemas/user.json" <- injection
                      }
                    },
                    "pagination": {
                      "$ref": "##/components/schemas/Pagination" <- local reference
                    }
                  }
                }
              }
            }
          ]
        },
        "example": {
          "$ref": "example.200.json" <- injection
        }
      }
    }
  }
}
```

One of the things I feel ColdBox developers should use more is the OpenApi `components.schemas` definition section.  Not only does referencing existing models create smaller (and faster) swagger files, but it also mirrors the same development approach taught in the Ortus API master class where the models for the API live in the app's root (e.g. models/User.cfc), and the API REST handlers live in sub-modules (e.g. /api/v1/users).  Defining models/schemas in the swagger root feels like a best practice to me and I believe someday I can come up with a way to dynamically inject schemas into this section. The below snippet example is taken from the root coldboxRest.json file:

```
  "components": {
     "schemas": {
            "Pagination": { "$ref": "_schemas/pagination.json" },
            "User": { "$ref": "paths/users/_schemas/user.json" },
            "Post": { "$ref": "paths/posts/_schemas/post.json" },
            "Media": { "$ref": "paths/media/_schemas/media.json" },
            "BookMedia": { "$ref": "paths/media/_schemas/bookMedia.json" },
            "MusicMedia": { "$ref": "paths/media/_schemas/musicMedia.json" }
        }
    },
    "paths": {
        "$extend": [
            { "$ref": "paths/users/users.json" },
            { "$ref": "paths/posts/posts.json" },
            { "$ref": "paths/media/media.json" }
        ]
    }
```
Some people may argue that the root app shouldn't have any knowledge about anything going on in submodules. I agree with that statement, except that the root app still needs to know about dependencies (hence the box.json file), so I don't see why the root of the swagger can't list out schema dependencies in the same way.  Either way, with this patch, Coldbox developers should have a lot more flexibility with how they document their APIs.

One additional advantage of defining models in the `components.schemas` section is that UI tools that render swagger files can display all the schemas in a special easy-to-read section like this:

![image](https://user-images.githubusercontent.com/5341142/204676788-21d6bf48-119d-4efa-82bc-d0a66964e1b0.png)

Anyway, that's it! Hopefully, you feel these contributions have a positive impact on the project. :)

homestar9 committed:
Add support for $extend to merge JSON files

Deprecated $allOf and $oneOf for performance reasons and to eliminate confusion with OpenApi keys 'allOf' and 'oneOf' with the same name.

homestar9 committed:
Bubble up swagger exceptions for better debugging

The parser works recursively, so if there's an exception thrown somewhere deep in the document structure, we want to display that error instead of the top-most exception which won't be as helpful.

homestar9 committed:
Better handling of $ref statements with ##

When a double pound (##) $ref statement is encountered, the user intends to literally reference something within the swagger doc.  Both the $ref and # reference should be returned.

homestar9 committed:
Add Coldbox RESTful tests


Removed debugging code

homestar9 committed:
Updated parser spec to check for $extend instead of $allOf

homestar9 committed:
Rename dyanamic composition keys $allOf, $oneOf to $extend

Performance boost and also avoids confusion with OpenApi keys named "allOf" and "oneOf" which are used in schema composition.

homestar9 committed:
Fix parser recursion when fetching document references

homestar9 commented:
Not sure why the ACF2021 test is failing. 
Edit: NVM found it.

homestar9 committed:
Removed extra comma

homestar9 commented:
@jclausen I discovered another possible recursion issue (not related to this PR) while writing some additional tests.

If you have a `$ref` that loads a file with `$extend` (formerly `$allOf`), the `$ref` comes back as a `Parser` instead of a struct.  Therefore we can never evaluate the `$extend`.  I created a fix in development by adding the following lines to the `parseDocumentInheritance()` method:
https://gist.github.com/homestar9/346110b21dd7af27213e1a5fcd294377#file-parser-cfc-L30-L36

That being said, we now have two separate checks for nested `Parser` objects, which adds redundancy, IMO.  I believe it would be better if any time a `$ref` is retrieved, it is normalized first, so it always comes back as a struct.  I don't know why we would ever need a nested instance of `Parser`.  Isn't the top-level `Parser` satisfactory?

homestar9 committed:
Bugfix: Nested Parser objects must be normalized

$ref contains $extend

homestar9 committed:
Fix issue with double pound signs at root

Also added tests for more complex/nested scenarios.

homestar9 committed:
Remove debugging code

jclausen commented:
@homestar9 Sorry for the delay on evaluating this.  I was travelling all week last week.  I have two concerns with the choice of `##`:  

1.  In a handler, this means you are going to have to escape it with `####`  which seems klunky
2.  Since we no ignore references we can't find, and return back the original `$ref` pointer, is it necessary to have the convention any longer?  

Other than that, I think this is a great improvement.

homestar9 committed:
Fixed issue with recursive $extend parsing

Nested $extend keys weren't being parsed. Wrote some additional tests and added some complexity to the coldbox json files to account for deep nesting.

homestar9 commented:
@jclausen, no problem at all. I've been deep diving into the Parser and looking for ways to make some improvements and tweaks geared toward Coldbox specific use-cases.

I agree with you. The `##` references are a code smell, and the more I've been playing around with it, the less I like it. We may be going about dynamically processing `$ref` keys incorrectly.  We currently have two intents when it comes to `$ref` keys in a swagger file:

- Point to an internal reference within the swagger file (e.g. `{ "$ref": "##/components/schemas/Media" }`
- Include the contents of an external .json file or .json key  (e.g. { "$ref": "../_schemas/media.json" }` or { "$ref": "#/components/schemas/Media" }

The OpenApi specs already dictate `$ref` keys should point to an address within the current Swagger file.  To avoid confusion between the OpenApi spec, I'm leaning toward the idea of using a new key called `$inject`, which would handle the second intent we currently use to include/inject the contents from an internal or external .json file.

I'm still playing around with the idea but here's an example of how the change might look:
```
// original paths json for a handler
{
    "/users": {
        "get": {
            "summary": "Paginate users",
            "operationId": "users.index",
            "tags": ["users"],
            "parameters": {
                "$ref": "index/parameters.json#/parameters"
            },
            "responses": {
                "$ref": "index/responses.json"
            }
        }
    },
    "/users/{id}": {
        "get": {
            "summary": "Show a user",
            "operationId": "users.show",
            "tags": ["users"],
            "parameters": {
                "$ref": "show/parameters.json#/parameters"
            },
            "responses": {
                "$ref": "show/responses.json"
            }
        }
    }
}
```

```
// revised $inject key
{
    "/users": {
        "get": {
            "summary": "Paginate users",
            "operationId": "users.index",
            "tags": ["users"],
            "parameters": {
                "$inject": "index/parameters.json#/parameters"
            },
            "responses": {
                "$inject": "index/responses.json"
            }
        }
    },
    "/users/{id}": {
        "get": {
            "summary": "Show a user",
            "operationId": "users.show",
            "tags": ["users"],
            "parameters": {
                "$inject": "show/parameters.json#/parameters"
            },
            "responses": {
                "$inject": "show/responses.json"
            }
        }
    }
}
```

One of the big advantages of changing from the `$ref` key to `$inject` is that it disambiguates the intent of the key.  In other words, we don't have to wonder if the user wants to reference an existing document key or inject some content into the structure.  At the very least, it makes reading and testing the JSON files easier.

What do you think?  If I have time, I may create a new branch to start testing it.

homestar9 commented:
I believe we need to change this test in ParserSpec.cfc.  Because we need to recurse the document for nested `$extend` references, we normalize child OpenApiDocument objects:

```
it( "Tests for the recursive presence of OpenAPIDocument objects within Parser #Parser.getSchemaType()# document object" , function(){
		var ParserDoc = Parser.getDocumentObject();
		expect( ParserDoc ).toBeInstanceOf( "Document" );
		expect( ParserDoc ).toHaveKey( "getDocument" );
		var APIDoc = ParserDoc.getDocument();
		expect( APIDoc ).toBeStruct();
		expect( APIDoc ).toHaveKey( "paths" );
		expect( APIDoc.paths ).toBeInstanceOf( "Parser" ); <-- should no longer contain any nested Parser objects
		runParserTypeChecks( APIDoc.paths );
		var paths = APIDoc.paths.getDocumentObject().getNormalizedDocument();
		expect( paths ).toHaveKey( "/pets" );
});
```

homestar9 committed:
Updated test expectation

Because we look for $extend keys recursively through the document object, we can no longer have nested Parser objects.  We shouldn't need them anyway. The top level Parser should be enough.

homestar9 commented:
Adjusted the test spec accordingly in the latest commit.

jclausen commented:
Sorry this one got lost in the weeds @homestar9 .  Thank you for all of your hard work on this.  I will be planning on bumping a release to this and `cbSwagger` next week
